source/2019-06-16-ember-3-11-released.md

```diff
@@ -0,0 +1,177 @@
+---
+title: Ember 3.11 Released
+author: Kenneth Larsen
+tags: Releases, 2019, 3, 3.11
+responsive: true
+---
+
+Today the Ember project is releasing version 3.11 of Ember.js, Ember Data, and Ember CLI. This release kicks off the 3.12 beta cycle for all sub-projects. We encourage our community (especially addon authors) to help test these beta builds and report any bugs before they are published as a final release in six weeks' time. The [ember-try](https://github.com/ember-cli/ember-try) addon is a great way to continuously test your projects against the latest Ember releases.
+
+You can read more about our general release process here:
+
+- [Release Dashboard](http://emberjs.com/builds/)
+- [The Ember Release Cycle](http://emberjs.com/blog/2013/09/06/new-ember-release-process.html)
+- [The Ember Project](http://emberjs.com/blog/2015/06/16/ember-project-at-2-0.html)
+- [Ember LTS Releases](http://emberjs.com/blog/2016/02/25/announcing-embers-first-lts.html)
+
+---
+
+## Ember.js
+
+Ember.js is the core framework for building ambitious web applications.
+
+### Changes in Ember.js 3.11
+Ember.js 3.11 is an incremental, backwards compatible release of Ember with bugfixes, performance improvements, and minor deprecations. There is COUNT (#) new feature, COUNT (#) deprecations, and COUNT (#) bugfixes in this version.
+
+#### New Features (4)
+
+Forwarding Element Modifiers with "Splattributes" (1 of 4)
+
+Angle bracket component invocation was introduced in Ember.js 3.7. Aside from the syntatic differences, the angle bracket invocation syntax enabled passing HTML attributes to components, which can then be applied the underlying HTML element(s) in the component's layout using the ...attributes "splattributes" syntax.
+
+This features clarifies how the "splattributes" feature interact with element modifiers.
+
+For more information please refer to [the RFC](https://github.com/emberjs/rfcs/blob/master/text/0435-modifier-splattributes.md).
+
+The {{fn}} helper (2 of 4)
+
+This new helper aims to remove some of the complexity of the current `action` helper. Some of these complex behaviors are:
+
+1. Argument partial application (currying)
+2. `this` context binding
+3. `send` checks for Component and Controllers
+
+The `fn` helper will take in a function and then the set of arguments that will be partially applied to the function.
+
+Here's a simple case on argument curry:
+
+```hbs
+{{fn this.log 1}}
+```
+
+```js
+return function() {
+  this.log.call(this, 1);
+}
+```
+
```

Comment:
Some suggestions for this section. My coworkers were confused (reading the CHANGELOG/RFC, not the blog post) about what this feature is for, so I spent some time thinking about how to explain it today.

How about:

> The `{{fn}}` helper provides a way to pass arguments to actions, particularly with native classes and the `@action` decorator.
>
> With the `@action` decorator, it is very straightforward to pass actions to other components ("data-down, actions-up"):
> 
> ```js
> import Component from "@ember/component";
> 
> class CounterComponent extends Component {
>   count = 0;
>   
>   @action
>   countUp(increment = 1) {
>     this.incrementProperty("count", increment);
>   }
> }
> ```
> 
> ```hbs
> Current count: {{this.count}}
> <MyButton @click={{this.countUp}}>Add One</MyButton>
> ```
> 
> The `@action` decorator ensures that the `this` value in the `countUp` action will always be the component instance, which allows it to be freely passed to other components. Where this falls short is when we need to pass arguments to these actions.
> 
> The `{{fn}}` helper provides a way to pass arguments into actions and "bundle" them up, so that they can be passed around to other components and still retain the provided arguments when called:
> 
> ```hbs
> Current count: {{this.count}}
> 
> {{!-- when not passing arguments, these are equivalent --}}
> <MyButton @click={{this.countUp}}>Add One</MyButton>
> <MyButton @click={{fn this.countUp}}>Add One</MyButton>
> 
> {{!-- calls this.countUp(10) when clicked --}}
> <MyButton @click={{fn this.countUp 10}}>Add Ten</MyButton>
> ```
> 
> In addition to the basic use case shown here, the `{{fn}}` helper supports other advanced use cases, such as adding more arguments to an existing function.
> 
> It should also be noted that the `{{action}}` helper can previously be used to accomplish similar functionalities, but due to some historical decisions, it may produce surprising results in some cases. Therefore, Ember users are encouraged to migrate to the `{{fn}}` helper along with the `@action` decorators where possible and appropriate. Refer to [the RFC](https://github.com/emberjs/rfcs/blob/master/text/0470-fn-helper.md#detailed-design) for more details and examples.